src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,1519 @@
+			}
+			$this->release_bookmark('internal_outer_html');
+		}
+	}
+
+	private function matching_closer() {
+		if($this->is_tag_closer()) {
+			return false;
+		}
+
+		$depth = $this->depth();
+		$token = $this->current_token;
+		while($this->process_next_tag()) {
+			if(
+				// Current element popped off the stack
+				$this->depth() <= $depth 
+				&& end($this->open_elements) !== $token
+			) {
+				if ($this->is_tag_closer() && $this->get_tag() === $token->tag) {
+					return array(
+						'closer_starts_at' => $this->tag_starts_at(),
+						'closer_ends_at' => $this->tag_ends_at,
+					);
+				} else {
+					return array(
+						'closer_starts_at' => $this->tag_starts_at(),
+						'closer_ends_at' => $this->tag_starts_at() - 1,
+					);
+				}
+			}
+		}
+		return array(
+			'closer_starts_at' => strlen($this->html),
+			'closer_ends_at' => strlen($this->html) - 1,
+		);
+	}
+
+	public function next_node() {
+		while ($this->process_next_tag()) {
+			// is_tag_closer can be NULL if `next_tag`
+			// didn't find a tag closer
+			if (false === $this->is_tag_closer()) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private $is_closing_open_tags = false;
+	private function process_next_tag() {
+		/*
+		 * We're done with the document but some tags
+		 * are still open. Let's close them one at a time.
+		 */
+		if ( $this->is_closing_open_tags ) {
+			// If only the root element is open, we're done.
+			if(count($this->open_elements) <= 1)
+			{
+				return false;
+			}
+
+			// Otherwise close the next open tag on the stack
+			$this->current_token = null;
+			$this->current_token_start = strlen($this->html);
+			$this->current_token_end = strlen($this->html);
+
+			$this->pop_open_element();
+			$this->get_updated_html();
+
+			parent::next_tag(array('tag_closers' => 'visit'));
+			$this->current_token = new WP_HTML_Tag_Token($this->get_tag());
+			$this->current_token_start = $this->tag_name_starts_at - 2;
+			$this->current_token_end = $this->tag_ends_at;
+			return true;
+		}
+		
+		/**
+		 * Go to the next tag and process any text was found along the way.
+		 */
+		$text_start = $this->tag_ends_at + 1;
+		if (!parent::next_tag(array('tag_closers' => 'visit'))) {
+			$this->process_text($text_start, strlen($this->html));
+
+			$this->is_closing_open_tags = true;
+			return $this->process_next_tag();
+		}
+
+		/**
+		 * We found a tag! Let's process any text we may have found along the way.
+		 */
+		$current_tag_start = $this->tag_starts_at();
+		$this->process_text($text_start, $current_tag_start);
+
+		$this->current_token = new WP_HTML_Tag_Token($this->get_tag());
+		$this->current_token_start = $current_tag_start;
+		$this->current_token_end = $this->tag_ends_at;
+
+		if ( ! $this->is_tag_closer() ) {
+			dbg( "Found {$this->current_token->tag} tag opener" );
+			switch ( $this->current_token->tag ) {
+				case 'HTML':
+					$this->drop_current_tag_token();
```

Comment:
what's the purpose here? it seems like we're modifying a document that nobody asked to modify

Reply:
This specific case is a hack I added to prevent a failure upon finding an `HTML` tag. It shouldn't be needed anymore once we look into other insertion modes. Now that I think about it, perhaps a `return` statement to ignore it would be a better hack.

Reply:
As for the general `drop_current_tag_token` function, it used to be `ignore_token`. Let's make it `ignore_token` again – there's no need to make such uninvited updates.